packages/apify-extra/src/dataset.ts

```diff
@@ -0,0 +1,339 @@
+            }
+            // We get the number of items first and then we precreate request info objects
+            let datasetInfo: Dataset | undefined;
+            if (loadFromLocalDataset) {
+                const dataset = await Actor.openDataset(datasetId);
+                datasetInfo = await dataset.getInfo() as Dataset;
+            } else {
+                datasetInfo = await client.dataset(datasetId).get();
+            }
+            if (!datasetInfo) {
+                throw new Error(`${LOG_PREFIX} Dataset ${datasetId} was not found`);
+            }
+            const { itemCount } = datasetInfo;
+            if (debugLog) {
+                log.info(`Dataset ${datasetId} has ${itemCount} items`);
+            }
+            const numberOfBatches = Math.ceil(itemCount / batchSize);
+
+            for (let i = 0; i < numberOfBatches; i++) {
+                const localOffsetLimit = calculateLocalOffsetLimit({ offset, limit, localStart: i * batchSize, batchSize });
+                if (!localOffsetLimit) {
+                    continue;
+                }
+
+                if (processFnLoadingState) {
+                    if (!processFnLoadingState[datasetId][localOffsetLimit.offset]) {
+                        processFnLoadingState[datasetId][localOffsetLimit.offset] = { done: false };
+                    } else if (processFnLoadingState[datasetId][localOffsetLimit.offset].done) {
+                        if (debugLog) {
+                            log.info(`Batch for dataset ${datasetId}, offset: ${localOffsetLimit.offset} was already processed, skipping...`);
+                        }
+                        continue;
+                    }
+                }
+
+                requestInfoArr.push({
+                    index: i,
+                    offset: localOffsetLimit.offset,
+                    limit: localOffsetLimit.limit,
+                    datasetId,
+                    datasetIndex,
+                });
+            }
+
+            datasetIndex++;
+        }
+        return requestInfoArr;
+    };
+
+    // This is array of arrays. Top level array is for each dataset and inside one entry for each batch (in order)
+    const loadedBatchedArr: DatasetItem[][][] = [];
+
+    let totalLoaded = 0;
+    const totalLoadedPerDataset: Record<string, number> = {};
+
+    const processFnLoadingState: any = persistLoadingStateForProcesFn
+        ? (await Actor.getValue(PROCESS_FN_LOADING_STATE_KV_RECORD_KEY) || {})
+        : null;
+
+    Actor.on('persistState', async () => {
+        await Actor.setValue(PROCESS_FN_LOADING_STATE_KV_RECORD_KEY, processFnLoadingState);
+    });
+
+    const requestInfoArr = await createRequestArray(processFnLoadingState);
+    if (debugLog) {
+        log.info(`Number of requests to do: ${requestInfoArr.length}`);
+    }
+
+    //  Now we execute all the requests in parallel (with defined concurrency)
+    await bluebird.map(requestInfoArr, async (requestInfoObj) => {
```

Comment:
should be the same as:

```suggestion
    await Promise.all(requestInfoArr.map(async (requestInfoObj) => {
```

and we can let bluebird rest in its graveyard

Reply:
Nope, it is there for the concurrency option. I can replace it later with AutoscaledPool, BasicCrawler or maybe I will find something more lightweight

Reply:
[something more lightweight?](https://github.com/apify/apify-sdk-js/pull/81/commits/f156a7f67357b53aea4b1da9b84cccf93d475f42#diff-f2f82b79930ea9b421746fa98f15f05cdd0c98f879ed1116dd1dfafb69740783R9)